### DISCUSSION TITLE
Allow _.Property shorthand for lambda functions

### DISCUSSION THREAD

**Discussion by lucasteles** [dotnet/csharplang#7607] [Language Ideas] [answered]
This is a feature that is coming in the next F# release https://github.com/fsharp/fslang-suggestions/issues/506 that I believe would be great to have in C#

It's very usual to have `Func<>` or `Expression<Func<>>` with just Property mapping, like mapping, LINQ, EF Model configuration, etc.

The idea is to have a syntax sugar for this kind of common definition. Probably replacing `x => x.Foo` to `_.Foo` 

LINQ example:
```csharp
using System.Linq;

Person[] persons = [];

// normal
var result1 = persons.Select(p => p.Name).GroupBy(p => p.Length);

// propose _
var result2 = persons.Select(_.Name).GroupBy(_.Length);

// propose just "."
var result3 = persons.Select(.Name).GroupBy(.Length);

record Person(string Name);

```

**Comment by En3Tho** [comment]
While I like F# I really feel like using `_` is a bad choice honestly. `_` is used as discard in both languages and now suddenly in F# it is a "valid" variable name if only as a shorthand. I personally like something like Kotlin's `it` much more - it's a proper variable name.

Even `.` feel like a better choice than `_`, although I don't really like it either for similar reasons.

Also don't forget that in F# you have to write `fun` keyword for lamdas and you can't omit it. C#'s lambdas are less verbose by design.

```fsharp
let result1 = persons.Select(fun p -> p.Name).GroupBy(fun p -> p.Length)
```

```csharp
var result1 = persons.Select(p => p.Name).GroupBy(p => p.Length);
```

**Reply by AndersKehlet** [reply]
I agree with this. I'd really prefer an implicit 'it'. 
I don't think property access alone would justify this feature. Cribbing Kotlin's syntax for this specific case doesn't feel like it would be too bad, though.

`persons.Select({ it.Name });`

Curlies already indicate scope in C#.

**Reply by lucasteles** [reply]
Just a note, F# already uses `_` for placeholding, discarding is just one case of usage

for example, `_` can be used for partial type inference:
```fs
let a1: List<int> = [1;2;3]
let a2: List<_> = [1;2;3]
let a3: int list = [1;2;3]
let a4: _ list = [1;2;3]
let a5: _ = [1;2;3]
let a6 = a1 :> IEnumerable<_>
```

**Reply by smoothdeveloper** [reply]
`_` is discard, only when it is free standing, so long it is followed by `.` it becomes something else, so long it is used in a type annotation, it is something else, etc.

`it` (and `_`) are valid identifiers in C#, I think it would make sense to be consistent with F#, since "it got it first" (or "_ got _ first").

IMO, in shaping the ecosystem, it is better to C# more consistent with F#, for small nudges like this, than with Kotlyn (no matter C# looks more like kotlyn than OCaml, for better or worse).

In F#, the feature allows for property and other type of member access too, I think this suggestion should reflect this as well.

**Reply by smoothdeveloper** [reply]
@AndersKehlet, is there a syntax for lambda, that uses curly, and no semicolon, to return a value like you show?

Why do we need to make C# look more like kotlyn, if `persons.Select(_.Name)` which is more readable, and wouldn't introduce new ways to read braces (that are taking significant share of C# code already, not always for best, such as brace style wars, etc.) would do?

**Comment by HaloFour** [comment, marked as answer]
See: #91

**Comment by saint4eva** [comment]
This is not needed in anyway.

**Reply by smoothdeveloper** [reply]
#91 seems to be "wanted", I don't know for sure all language features met with a shipping implementation based on "absolute need in some way" only.

Your statement could apply to many syntax sugar C# added since V1.0.

**Reply by saint4eva** [reply]
This language feature request will not add any value to the ecosystem.

**Reply by smoothdeveloper** [reply]
Going there, I can say the same for many C# features added. #91 has many upvotes anyways.

Maybe more accurate for you to express you don't like or support the addition, which I think is fair to share.

**Reply by saint4eva** [reply]
As I said, what the ecosystem already has is better and it is working greatly.
